open-machine-learning-jupyter-book/deep-learning/CNN.md

```diff
@@ -13,4 +13,1201 @@ kernelspec:
+
+def render_deepdream(t_obj, img0=img_noise,
+                     iter_n=10, step=1.5, octave_n=4, octave_scale=1.4):
+    # defining the optimization objective, the objective is the mean of the feature
+    t_score = tf.reduce_mean(t_obj)
+    # Our gradients will be defined as changing the t_input to get closer to
+    # the values of t_score.  Here, t_score is the mean of the feature we select,
+    # and t_input will be the image octave (starting with the last)
+    t_grad = tf.gradients(t_score, t_input)[0] # behold the power of automatic differentiation!
+
+    # Store the image
+    img = img0
+    # Initialize the octave list
+    octaves = []
+    # Since we stored the image, we need to only calculate n-1 octaves
+    for i in range(octave_n-1):
+        # Extract the image shape
+        hw = img.shape[:2]
+        # Resize the image, scale by the octave_scale (resize by linear interpolation)
+        lo = resize(img, np.int32(np.float32(hw)/octave_scale))
+        # Residual is hi.  Where residual = image - (Resize lo to be hw-shape)
+        hi = img-resize(lo, hw)
+        # Save the lo image for re-iterating
+        img = lo
+        # Save the extracted hi-image
+        octaves.append(hi)
+    
+    # generate details octave by octave
+    for octave in range(octave_n):
+        if octave>0:
+            # Start with the last octave
+            hi = octaves[-octave]
+            #
+            img = resize(img, hi.shape[:2])+hi
+        for i in range(iter_n):
+            # Calculate gradient of the image.
+            g = calc_grad_tiled(img, t_grad)
+            # Ideally, we would just add the gradient, g, but
+            # we want do a forward step size of it ('step'),
+            # and divide it by the avg. norm of the gradient, so
+            # we are adding a gradient of a certain size each step.
+            # Also, to make sure we aren't dividing by zero, we add 1e-7.
+            img += g*(step / (np.abs(g).mean()+1e-7))
+            print('.',end = ' ')
+        showarray(img/255.0)
+
+# Run Deep Dream
+if __name__=="__main__":
+    # Create resize function that has a wrapper that creates specified placeholder types
+    resize = tffunc(np.float32, np.int32)(resize)
+    
+    # Open image
+    img0 = PIL.Image.open('book_cover.jpg')
+    img0 = np.float32(img0)
+    # Show Original Image
+    showarray(img0/255.0)
+
+    # Create deep dream
+    render_deepdream(T(layer)[:, :, :, channel], img0, iter_n=15)
+
+    sess.close()
+```
+
+
+---
+
+## Your turn! 🚀
+
+Practice the Convolutional Neural Networks by following this xxxx.
+
+## Self study
+
+You can refer to those YouTube videos for further study:
+- [Convolutional Neural Networks (CNNs) explained, by deeplizard](https://www.youtube.com/watch?v=YRhxdVk_sIs)
+- [Convolutional Neural Networks Explained (CNN Visualized), by Futurology](https://www.youtube.com/watch?v=pj9-rr1wDhM)
+
+
+## [optional] At the frontier of the research world

```

Comment:
We don't have such section. You could make it as 3rd level header as part of self study

Reply:
You can use 'Research trend' as the title? Or any other better name.